File: PHP/7/example7.php
#!/usr/bin/env php

<?php

function do_something() {
    echo "I am doing something.\n";
}


function do_something_with($num) {
    echo "I am doing somethig with " . $num . "\n";
}


function do_something_and_return($num) {
    return $num * 2;
}


function do_something_with_opt($num, $num2=null)
{
    if ($num2) {
        return $num * $num2;
    }
    return $num * 2;
}


function do_something_with_args($num, $num2=null)
{
    echo "Parameter: " . $num . "\n";
    if ($num) {
        echo "Parameter: " . $num2 . "\n";
    }
    echo "Parameter: " . func_get_arg(2) . "\n";
    echo "Parameter: " . func_get_arg(3) . "\n";
    echo "Parameter: " . func_get_arg(4) . "\n";
}


function do_something_plus($options)
{
    echo "Lucky number: " . $options["lucky"] . "\n";
    echo "Unlucky number: " . $options["unlucky"] . "\n";
}


echo "\nCalling doSomething() ...\n";
do_something();

echo "\nCalling do_something_with() ...\n";
do_something_with(7);

echo "\nCalling do_something_and_return() ...\n";
echo do_something_and_return(7) . "\n";

echo "\nCalling do_something_with_opt() ...\n";
echo do_something_with_opt(7) . "\n";
echo do_something_with_opt(7, 4) . "\n";

echo "\nCalling do_something_with_args() ...\n";
do_something_with_args(5, $num2=10, "eenie", "minie", "mo");
do_something_with_args(5, "eenie", "minie", "mo", "moo");

$extra_nums = ["lucky" => 7, "unlucky" => 13];
echo "\nCalling do_something_plus() ...\n";
do_something_plus($extra_nums);


?>

File: Java/basic/src/main/java/org/prodigius/javabasic/example7/Example7.java
package org.prodigius.javabasic.example7;

import java.util.Map;
import java.util.HashMap;

public class Example7
{
    public static void doSomething()
    {
        System.out.println("I am doing something.");
    }


    public static void doSomething(int num)
    {
        System.out.println("I am doing something with " + num);
    }


    public static void doSomethingWith(int num)
    {
        System.out.println("I am doing something with " + num);
    }


    public static int doSomethingAndReturn(int num)
    {
        return num * 2;
    }


    public static int[] doSomethingWithOptions(int num, int... nums)
    {
        int[] results;

        if (nums.length == 0)
        {
            results = new int[1];
            results[0] = num * 2;
        }
        else
        {
            results = new int[nums.length + 1];
            results[0] = num * 2;
            for (int i=1; i < results.length; i++)
            {
                results[i] = nums[i - 1] * 2;
            }
        }

        return results;
    }


    public static int[] doSomethingPlus(
        int num, Map<String,Integer> extraNums, int... nums)
    {

        int[] res = doSomethingWithOptions(num, nums);
        for (int n : res)
        {
            System.out.println(n);
        }

        int[] results = new int[res.length + extraNums.size()];

        int i;
        for (i=0; i < res.length; i++)
        {
            results[i] = res[i];
        }

        for (int v : extraNums.values())
        {
            results[i] = v;
            i++;
        }

        return results;

    }

    public static void main(String[] args)
    {
        System.out.println("Calling doSomething() ...");
        doSomething();

        System.out.println("Calling doSomething(7) ...");
        doSomethingWith(7);
        doSomething(7);

        System.out.println("Calling doSomethingAndReturn(7) ...");
        System.out.println(doSomethingAndReturn(7));

        System.out.println("Calling doSomethingWithOptions(7) ...");

        for (int n : doSomethingWithOptions(7))
        {
            System.out.print(n + " .. ");
        }

        for (int n : doSomethingWithOptions(7, 9, 11, 13))
        {
            System.out.print(n + " .. ");
        }

        System.out.println("");

        Map<String, Integer> extraNums = new HashMap<String, Integer>();
        extraNums.put("lucky", 7);
        extraNums.put("unlucky", 13);

        System.out.println("Calling doSomethingPlus() ...");

        for (int n : doSomethingPlus(7, extraNums, 9, 11, 13))
        {
            System.out.print(n + " .. ");
        }

        System.out.println("");
    }
}